library/core/src/option.rs

```diff
@@ -597,6 +597,30 @@ impl<T> Option<T> {
         !self.is_some()
     }
 
+    /// Returns `true` if the option is a [`None`] or the [`Some`] value
+    /// inside of it matches a predicate.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(is_some_with)]
+    ///
+    /// let x: Option<u32> = Some(2);
+    /// assert_eq!(x.is_none_or(|&x| x > 1), true);
+    ///
+    /// let x: Option<u32> = Some(0);
+    /// assert_eq!(x.is_none_or(|&x| x > 1), false);
+    ///
+    /// let x: Option<u32> = None;
+    /// assert_eq!(x.is_none_or(|&x| x > 1), true);
+    /// ```
+    #[must_use]
+    #[inline]
+    #[unstable(feature = "is_some_with", issue = "93050")]
+    pub fn is_none_or(&self, f: impl FnOnce(&T) -> bool) -> bool {
+        self.is_none() || self.is_some_and(f)
```

Comment:
Do we actually want a combinator function for this, rather than just inlining this pair of calls?

`is_some_and` expresses something not easily written another way. I'm not sure that justification applies to `is_none_or`.

Reply:
Good question! There was [some talk in the issue](https://github.com/rust-lang/rust/issues/93050#issuecomment-1177890673) that `is_none_or` would actually be *more* justified than `is_some_and` since the former isn't easily expressible through `if let` or `matches!`. Also, I had previously come to the conclusion that [there was not a great way to express this idiom](https://github.com/rust-lang/rust/issues/93050#issuecomment-1215425684), but I had not considered using `is_some_and` itself to calculate it, and I actually find this expression quite readable. My main argument in favor of `is_none_or` is the reduced cognitive load, as well as just plain symmetry with `is_{some,ok,err}_and`.

Reply:
Oh interesting. I also hadn't fully realized yet that `is_none_or` can be written as simply `is_none || is_some_and`.

Reply:
`is_none || is_some_and` is redundant since you are checking the variant twice. `match` would be more appropriate here IMO.

Reply:
Yep, you were right. It's in the latest commit.

Reply:
I have found that writing `x.is_none() || x.is_some_and(...)` is not always as ergonomic as `is_none_or(...)` because it forces you to store the `Option` in a variable first. (And, frankly, periodically I keep forgetting which of the two was added and which of the two wasn't)